File: stack/1190_reverse_substrings_between_each_pair_of_parentheses/1190_reverse_substrings_between_each_pair_of_parentheses.go
package main

import "fmt"

func reverseParentheses(s string) string {
	var res []byte
	for i := 0; i < len(s); i++ {
		if s[i] != ')' {
			res = append(res, s[i])
		} else {
			var temp []byte
			for res[len(res)-1] != '(' {
				temp = append(temp, res[len(res)-1])
				res = res[:len(res)-1]
			}

			res = res[:len(res)-1]
			for j := 0; j < len(temp); j++ {
				res = append(res, temp[j])
			}
		}
	}

	return string(res)
}

func main() {
	s := "(abcd)"
	res := reverseParentheses(s)
	fmt.Println(res)
}

File: stack/1190_reverse_substrings_between_each_pair_of_parentheses/1190_reverse_substrings_between_each_pair_of_parentheses.py
class Solution:
    def reverseParentheses(self, s: str) -> str:
        stack = ['']
        for c in s:
            if c == '(':
                stack.append('')
            elif c == ')':
                reversed_top = stack.pop()[::-1]
                stack[-1] += reversed_top
            else:
                stack[-1] += c
            print(stack)
        return stack[0]


s = "(abcd)"
res = Solution().reverseParentheses(s)
print(res)